contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+        require(
+            _baseGenerationRate < boostedGenerationRate,
+            "VeJoeStaking: expected new _baseGenerationRate to be less than boostedGenerationRate"
+        );
+        baseGenerationRate = _baseGenerationRate;
+    }
+
+    /// @notice Set boostedGenerationRate
+    /// @param _boostedGenerationRate the new boostedGenerationRate
+    function setBoostedGenerationRate(uint256 _boostedGenerationRate) external onlyOwner {
+        require(
+            _boostedGenerationRate > baseGenerationRate,
+            "VeJoeStaking: expected new _boostedGenerationRate to be greater than baseGenerationRate"
+        );
+        boostedGenerationRate = _boostedGenerationRate;
+    }
+
+    /// @notice Set boostedThreshold
+    /// @param _boostedThreshold the new boostedThreshold
+    function setBoostedThreshold(uint256 _boostedThreshold) external onlyOwner {
+        require(
+            _boostedThreshold <= 100,
+            "VeJoeStaking: expected new _boostedThreshold to be less than or equal to 100"
+        );
+        boostedThreshold = _boostedThreshold;
+    }
+
+    /// @notice Set boostedDuration
+    /// @param _boostedDuration the new boostedDuration
+    function setBoostedDuration(uint256 _boostedDuration) external onlyOwner {
+        boostedDuration = _boostedDuration;
+    }
+
+    /// @notice Deposits JOE to start staking for veJOE. Note that any pending veJOE
+    /// will also be claimed in the process.
+    /// @param _amount the amount of JOE to deposit
+    function deposit(uint256 _amount) external {
+        require(_amount > 0, "VeJoeStaking: expected deposit amount to be greater than zero");
+
+        if (getUserHasNonZeroBalance(msg.sender)) {
+            // If user already has staked JOE, we first send them any pending veJOE
+            _claim();
+
+            userInfos[msg.sender].balance += _amount;
+
+            // User is eligible for boosted benefits if:
+            // 1. They are not already currently receiving boosted benefits
+            // 2. Their staked JOE is at least `boostedThreshold / 100 * totalStakedJoe`
+            if (userInfos[msg.sender].boostEndTimestamp == 0) {
+                uint256 totalStakedJoe = joe.balanceOf(address(this));
+                if ((userInfos[msg.sender].balance * 100) / boostedThreshold >= totalStakedJoe) {
+                    userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
+                }
+            }
+        } else {
+            // If the user's `lastRewardTimestamp` is 0, i.e. if this is the user's first time staking,
+            // then they will receive boosted benefits.
+            // Note that it is important we perform this check **before** we update the user's `lastRewardTimestamp`
+            // down below.
+            if (userInfos[msg.sender].lastRewardTimestamp == 0) {
+                userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
```

Comment:
Does this leave us open to an attack where a user can switch accounts and claim a boost for a minimum amount of joe? If I am a first time user I only need to deposit a minimum amount of joe to get a boost whereas non first time users need to hold 1% of the pool.

Reply:
See my comment above (https://github.com/traderjoe-xyz/joe-core/pull/29#discussion_r799954295)